barretenberg/cpp/pil/vm2/data_copy.pil

```diff
@@ -0,0 +1,175 @@
+
+    pol commit sel_end;
+    sel_end * (1 - sel_end) = 0;
+    pol WRITE_CNT_MINUS_ONE = data_copy_size - 1;
+    pol commit next_write_count_inv;
+    // sel_end = 1 IFF data_copy_size - 1 = 0;
+    #[END_WRITE_CONDITION]
+    SEL * (WRITE_CNT_MINUS_ONE * (sel_end * (1 - next_write_count_inv) + next_write_count_inv) - 1 + sel_end) = 0;
+
+    // ===== Writing to dst_context_id =====
+    pol commit one;
+    SEL * (1 - one) = 0;
+    // Data copy size decrements for each row until we end
+    #[DECR_COPY_SIZE]
+    SEL * (1 - sel_end) * (data_copy_size' - data_copy_size + 1) = 0;
+
+    pol commit sel_mem_write;
+    // Regardless of operation, we have to write
+    // TODO check out of bound
+    sel_mem_write = SEL;
+    #[INCR_WRITE_ADDR]
+    (1 - precomputed.first_row) * SEL * (1 - sel_end) * (write_addr' - write_addr - 1) = 0;
+    #[MEM_WRITE]
+    sel_mem_write { clk, write_addr, value, precomputed.zero, one, dst_context_id }
+    in
+    memory.sel { memory.clk, memory.address, memory.value, memory.tag, memory.rw, memory.space_id };
+
+    // ===== Reading for nested call =====
+    // === Helper ===
+    // When calling calldatacopy, the src_context_id == parent_id
+    pol commit is_top_level; // == 1 iff parent_id == 0
+    is_top_level * (1 - is_top_level) = 0;
+    pol commit parent_id_inv; // For zero-check of has_parent_ctx
+    #[TOP_LEVEL_COND]
+    SEL * (src_context_id * (is_top_level * (1 - parent_id_inv) + parent_id_inv) - 1 + is_top_level) = 0;
+
+    // === Establishing the read_addr ===
+    // (1) The initial Read Addr is data_addr + data_offset
+    // (2) Subsequent read addrs are incremented by 1 unless this is a padding row
+    // (3) Error handling if read_addr is OOB
+    pol commit read_addr;  // The addr to start reading the data from
+    SEL * sel_start * (read_addr - data_addr - data_offset) = 0;
+    #[INCR_READ_ADDR]
+    SEL * (1 - padding) * (1 - sel_end) * (read_addr' - read_addr - 1) = 0;
+
+    // === Establishing number of reads ===
+    // Number of reads: min(data_offset + data_copy_size, data_size)
+    pol commit read_count;
+    // TODO: @sel_start read_count = min(data_offset + data_copy_size, data_size)
+    #[DECR_READ_COUNT]
+    SEL * (1 - padding) * (read_count' - read_count + 1) = 0; // Decrement
+    pol commit padding; // Padding = 1 if read_count = 0
+    pol commit read_count_inv;
+    #[PADDING_CONDITION]
+    SEL * (read_count * (padding * (1 - read_count_inv) + read_count_inv) - 1 + padding) = 0;
+
+    // Read from memory if we are not the top level call and not a padding row
+    pol commit sel_mem_read; // If the current row is a memory op read
+    sel_mem_read = SEL * (1 - is_top_level) * (1 - padding);
+
+    // === Value Padding ===
+    pol commit value;
+    #[PAD_VALUE]
+    SEL * padding * value = 0;
+
+    #[MEM_READ]
+    sel_mem_read { clk, read_addr, value, precomputed.zero, precomputed.zero, src_context_id }
```

Comment:
could you add comments in the `precomputed.zero`s? Like `/*rw=*/precomputed.zero/*(read)*/` and the other one is an FF?